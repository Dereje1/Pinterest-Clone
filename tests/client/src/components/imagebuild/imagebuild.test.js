import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import ImageBuild from '../../../../../client/src/components/imagebuild/imagebuild';
import { pinsStub } from '../../../pinsStub';

describe('The ImageBuild component', () => {
  let props;
  beforeEach(() => {
    props = {
      pinImage: jest.fn(),
      deletePin: jest.fn(),
      pinList: pinsStub,
      displayBrokenImage: false,
      ready: true,
    };
  });

  test('will render....', () => {
    const wrapper = shallow(<ImageBuild {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will call layoutComplete when masonry is done loading', () => {
    const wrapper = shallow(<ImageBuild {...props} />);
    let bubblecontainer = wrapper.find({ id: 'bubblecontainer' });
    expect(bubblecontainer.props().style).toEqual({ display: 'flex' });
    let masonry = wrapper.find('MasonryComponent');
    masonry.props().onImagesLoaded();
    bubblecontainer = wrapper.find({ id: 'bubblecontainer' });
    expect(bubblecontainer.props().style).toEqual({ display: 'none' });
    // re-fire to make sure that imagesloaded does not change state again
    masonry = wrapper.find('MasonryComponent');
    masonry.props().onImagesLoaded();
    bubblecontainer = wrapper.find({ id: 'bubblecontainer' });
    expect(bubblecontainer.props().style).toEqual({ display: 'none' });
  });

  test('will call pinEnlarge for a single pin', () => {
    const wrapper = shallow(<ImageBuild {...props} />);
    const pin = wrapper.find({ className: 'image-box' }).at(1);
    let pinZoom = wrapper.find('PinZoom');
    expect(pinZoom.isEmptyRender()).toBe(true);
    pin.props().onClick({ target: { type: 'any' }, pageY: 20, clientY: 10 });
    pinZoom = wrapper.find('PinZoom');
    expect(pinZoom.isEmptyRender()).toBe(false);
    expect(pinZoom.props().zoomInfo).toEqual([pinsStub[1], 10]);
  });

  test('will dismiss pinEnlarge for clicks near submit', () => {
    const wrapper = shallow(<ImageBuild {...props} />);
    const pin = wrapper.find({ className: 'image-box' }).at(1);
    let pinZoom = wrapper.find('PinZoom');
    expect(pinZoom.isEmptyRender()).toBe(true);
    pin.props().onClick({ target: { type: 'submit' }, pageY: 20, clientY: 10 });
    pinZoom = wrapper.find('PinZoom');
    expect(pinZoom.isEmptyRender()).toBe(true);
  });

  test('will dismiss pinEnlarge if pin already zoomed', () => {
    const wrapper = shallow(<ImageBuild {...props} />);
    let pin = wrapper.find({ className: 'image-box' }).at(1);
    let pinZoom = wrapper.find('PinZoom');
    expect(pinZoom.isEmptyRender()).toBe(true);
    pin.props().onClick({ target: { type: 'any' }, pageY: 20, clientY: 10 });
    pinZoom = wrapper.find('PinZoom');
    expect(pinZoom.isEmptyRender()).toBe(false);
    expect(pinZoom.props().zoomInfo).toEqual([pinsStub[1], 10]);
    // re-fire again after pin is already zooomed
    pin = wrapper.find({ className: 'image-box' }).at(1);
    pin.props().onClick({ target: { type: 'any' }, pageY: 20, clientY: 5 });
    pinZoom = wrapper.find('PinZoom');
    // If not working second arg should have been 15
    expect(pinZoom.props().zoomInfo).toEqual([pinsStub[1], 10]);
  });

  test('will reset pinEnlarge', () => {
    const wrapper = shallow(<ImageBuild {...props} />);
    const pin = wrapper.find({ className: 'image-box' }).at(1);
    let pinZoom = wrapper.find('PinZoom');
    expect(pinZoom.isEmptyRender()).toBe(true);
    pin.props().onClick({ target: { type: 'any' }, pageY: 20, clientY: 10 });
    pinZoom = wrapper.find('PinZoom');
    expect(pinZoom.isEmptyRender()).toBe(false);
    expect(pinZoom.props().zoomInfo).toEqual([pinsStub[1], 10]);
    pinZoom.props().reset();
    pinZoom = wrapper.find('PinZoom');
    expect(pinZoom.isEmptyRender()).toBe(true);
  });

  test('will handle onBrokenImage for an image load error on home page', () => {
    const wrapper = shallow(<ImageBuild {...props} />);
    let pin = wrapper.find({ className: 'image-box' }).at(1);
    let img = pin.find({ alt: 'imgDescription id-2' });
    img.props().onError();
    pin = wrapper.find({ className: 'image-box' }).at(1);
    img = pin.find({ alt: 'imgDescription id-2' });
    expect(img.isEmptyRender()).toBe(true);
  });

  test('will handle onBrokenImage for an image load error on profile page', () => {
    const updatedProps = {
      ...props,
      displayBrokenImage: true,
    };
    const wrapper = shallow(<ImageBuild {...updatedProps} />);
    let pin = wrapper.find({ className: 'image-box' }).at(1);
    let img = pin.find({ alt: 'imgDescription id-2' });
    img.props().onError();
    pin = wrapper.find({ className: 'image-box' }).at(1);
    img = pin.find({ alt: 'Broken Img - imgDescription id-2' });
    expect(img.isEmptyRender()).toBe(false);
  });

  test('will display the bubbles and hide images if not done loading', () => {
    const wrapper = shallow(<ImageBuild {...props} />);
    const bubbles = wrapper.find({ id: 'bubblecontainer' });
    const pin = wrapper.find({ className: 'image-box' }).at(1);
    const img = pin.find({ alt: 'imgDescription id-2' });
    expect(bubbles.props().style.display).toBe('flex');
    expect(img.props().style.visibility).toBe('hidden');
  });
});
