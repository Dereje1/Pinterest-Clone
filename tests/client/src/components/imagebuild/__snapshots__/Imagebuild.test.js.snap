// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ImageBuild component will render.... 1`] = `
<Fragment>
  <div
    id="mainframe"
  >
    <InfiniteScroll
      dataLength={3}
      hasMore={false}
      initialScrollY={0}
      next={[Function]}
    >
      <MasonryPins
        deletePin={[MockFunction]}
        layoutComplete={[Function]}
        onBrokenImage={[Function]}
        pinEnlarge={[Function]}
        pinImage={[MockFunction]}
        pins={
          Array [
            Object {
              "_id": 1,
              "comments": Array [
                Object {
                  "comment": "Test comment",
                  "createdAt": "2021-07-28T16:30:48.572Z",
                  "displayName": "tester-displayName",
                },
              ],
              "createdAt": "2020-12-25T00:00:00.000+00:00",
              "hasSaved": false,
              "imgDescription": "imgDescription id-1",
              "imgLink": "imgLink id-1",
              "owner": "owner id-1",
              "owns": false,
              "savedBy": Array [
                "savedBy - id-1",
                "savedBy2 - id-1",
              ],
            },
            Object {
              "_id": 2,
              "createdAt": "2020-12-25T00:00:00.000+00:00",
              "hasSaved": true,
              "imgDescription": "imgDescription id-2",
              "imgLink": "imgLink id-2",
              "owner": "owner id-2",
              "owns": false,
              "savedBy": Array [
                "savedBy - id-2",
                "savedBy2 - id-2",
              ],
            },
            Object {
              "_id": 3,
              "createdAt": "2020-12-25T00:00:00.000+00:00",
              "hasSaved": false,
              "imgDescription": "imgDescription id-3",
              "imgLink": "imgLink id-3",
              "owner": "owner id-3",
              "owns": true,
              "savedBy": Array [
                "savedBy - id-3",
                "savedBy2 - id-3",
              ],
            },
          ]
        }
      />
    </InfiniteScroll>
  </div>
  <Loading
    imagesLoaded={false}
    ready={true}
  />
</Fragment>
`;
